mitiq/zne.py

```diff
@@ -1,50 +1,122 @@
-from typing import List, Tuple, Callable
+"""Zero-noise extrapolation tools."""
+
+from typing import Callable
 import mitiq.qiskit.qiskit_utils as qs_utils
 from mitiq import QPROGRAM
 from mitiq.factories import Factory, LinearFactory
 
 
-def mitigate(pq: QPROGRAM, fac: Factory, scale_noise: Callable[[QPROGRAM, float], QPROGRAM], run_program: Callable) -> \
-        Tuple[List[float], List[float]]:
+def run_factory(fac: Factory, noise_to_expval: Callable[[float], float]) -> None:
     """
-    Runs the factory until convergence and returns the full list of noise values and expectations that have been
-    calculated.
-    :param pq: Program to mitigate.
+    Runs a factory until convergence and returns the final zero-noise extrapolation.
+
+    Args:
+        fac: Instance of Factory object to be run.
+        noise_to_expval: Function mapping noise scale values to expectation vales.
     """
-    if not fac.is_converged(fac.instack, fac.outstack):
-        next_param = fac.step(fac.instack, fac.outstack)
+    if not fac.is_converged():
+        next_param = fac.next()
+        next_result = noise_to_expval(next_param)
+        fac.push(next_param, next_result)
+
+        return run_factory(fac, noise_to_expval)
 
-        scaled_pq = scale_noise(pq, next_param)
+    return None
 
-        next_result = run_program(scaled_pq)
-        fac.outstack.append(next_result)
+# quantum version of run_factory. Similar to the old "mitigate".
+def qrun_factory(fac: Factory, qp: QPROGRAM, executor: Callable[[QPROGRAM], float], \
```

Comment:
This is I guess the equivalent of `test_factories.apply_algorithm` (see comment there). Can this be used in `test_factories.py`?

Reply:
Good point! I have just done as suggested.